File: daeun/stack_queue/프로세스.js
function solution(priorities, location) {
  var queue = priorities.map((priority, index) => ({
    index: index,
    priority: priority,
  }));

  var answer = 0;

  while (true) {
    var current = queue.shift();
    // 현재 프로세스보다 우선순위가 높은 프로세스가 있는지 확인
    // 우선순위 더 높은 프로세스 o -> 현재 프로세스를 queue의 맨 뒤로 옮김
    if (queue.some((process) => process.priority > current.priority)) {
      queue.push(current);
    } else {
      // 우선순위 더 높은 프로세스 x -> 현재 프로세스를 실행
      answer++;
      if (current.index === location) {
        return answer;
      }
    }
  }
}

File: Grace/stack_queue/프로세스.py
from collections import deque

def solution(priorities, location):
    answer = []
    queue = deque((i, j) for i, j in enumerate(priorities))
    
    while queue:
        process = queue.popleft()
        # 우선순위가 더 높은 프로세스가 하나라도 있으면 큐에 다시 추가 / 아니면 해당 프로세스 실행
        if queue and any(process[1] < q[1] for q in queue):
            queue.append(process)
        else:
            answer.append(process)
    
    # location 실행 순서 찾기
    for i in answer:
        if i[0] == location:
            return answer.index(i)+1